src/broadcast.jl

```diff
@@ -8,199 +9,92 @@ else
-        nargs = length(Xs)
-        shape = indices(Z)
-        check_broadcast_indices(shape, Xs...)
-        keeps, Idefaults = map_newindexer(shape, Xs)
-        Base.Broadcast._broadcast!(f, Z, keeps, Idefaults, Xs, Val{nargs}; lift=lift)
-        return Z
+Note that this method's signature specifies the source `As` arrays as all
+`NullableArray`s. Thus, calling `broadcast` on arguments consisting
+of both `Array`s and `NullableArray`s will fall back to the standard implementation
+of `broadcast` (i.e. without lifting).
+"""
+function Base.broadcast{F, N}(f::F, As::Vararg{NullableArray, N})
+    # These definitions are needed to avoid allocation due to splatting
+    f2(x1) = lift(f, x1)
+    f2(x1, x2) = lift(f, x1, x2)
+    f2(x1, x2, x3) = lift(f, x1, x2, x3)
+    f2(x1, x2, x3, x4) = lift(f, x1, x2, x3, x4)
+    f2(x1, x2, x3, x4, x5) = lift(f, x1, x2, x3, x4, x5)
+    f2(x1, x2, x3, x4, x5, x6) = lift(f, x1, x2, x3, x4, x5, x6)
+    f2(x1, x2, x3, x4, x5, x6, x7) = lift(f, x1, x2, x3, x4, x5, x6, x7)
+    f2(x...) = lift(f, x...)
+
+    T = _default_eltype(Base.Generator{ziptype(As...), ftype(f2, As...)})
+    if isleaftype(T) && !(T <: Nullable)
+        dest = similar(Array{eltype(T)}, broadcast_indices(As...))
+    else
+        dest = similar(NullableArray{eltype(T)}, broadcast_indices(As...))
     end
+    invoke_broadcast!(f2, dest, As...)
 end
 
 """
-    broadcast(f, As::NullableArray...;lift::Bool=false)
+    broadcast!(f, dest::NullableArray, As::NullableArray...)
 
-This method implements the same behavior as that of `broadcast` when called on
-regular `Array` arguments. It also includes the `lift` keyword argument, which
-when set to true will lift `f` over the entries of the `As`.
+Call `broadcast!` with nullable lifting semantics.
+Lifting means calling function `f` on the the values wrapped inside `Nullable` entries
+of the input arrays, and returning null if any entry is missing.
 
-Lifting is disabled by default. Note that this method's signature specifies the
-source `As` arrays as all `NullableArray`s. Thus, calling `broadcast!` on
-arguments consisting of both `Array`s and `NullableArray`s will fall back to the
-implementation of `broadcast` in `base/broadcast.jl`.
+Note that this method's signature specifies the destination `dest` array as well as the
+source `As` arrays as all `NullableArray`s. Thus, calling `broadcast!` on a arguments
+consisting of both `Array`s and `NullableArray`s will fall back to the standard implementation
+of `broadcast!` (i.e. without lifting).
 """
-@inline function Base.broadcast(f, Xs::NullableArray...;lift::Bool=false)
-    return broadcast!(f, NullableArray(eltype(promote_eltype(Xs...)),
-                                       _to_shape(broadcast_indices(Xs...))),
-                      Xs...; lift=lift)
+function Base.broadcast!{F, N}(f::F, dest::NullableArray, As::Vararg{NullableArray, N})
+    # These definitions are needed to avoid allocation due to splatting
```

Comment:
Have you verified that allocations are avoided for `f2` of more than two arguments? Also, did you try `@inline` decorations for `f2` to avoid the splatting penalty?